web/src/main/scala/i4s/symbolic/web/components/SentenceGraph.scala

```diff
@@ -59,33 +89,101 @@ import slinky.web.svg.{g, svg, className => svgClassName}
         .each { (t: SVGTextElement, d: JSTokenNode) =>
           d.width = t.getComputedTextLength() + padding
           d.offSet = accumulatedY
-          accumulatedY += d.width + margin
+          for(i <- d.edges.indices){
+            if(d.edges(i).target < d.position){
+              d.offSet = Math.max(d.offSet, graph.tokens(d.edges(i).target).offSet + d.edges(i).width)
+            }
+          }
+          accumulatedY = d.offSet + d.width + margin
           t.remove()
         }
 
+      for(edge <- edges){
+        edge.offSet = (edge.source.offSet + edge.source.width/2 + edge.targetRef.offSet + edge.targetRef.width/2)/2 - edge.width/2
+      }
+
       val selection = svg
         .select(".graph-area")
         .selectAll(".token")
-        .data(graph.tokens).enter()
+        .data(graph.tokens)
+        .enter()
         .append("g")
-        .attr("transform", (d: JSTokenNode, i: Int) => s"translate(${d.offSet},5)")
-
-        selection
-          .append("rect")
-          .attr("rx", 6)
-          .attr("ry", 6)
-          .attr("width", (d: JSTokenNode) => d.width)
-          .attr("height", 20)
-          .style("fill", "white")
-          .style("stroke", "gray")
-        selection
-          .append("text")
-          .text((d: JSTokenNode) => d.token)
-          .attr("x", (d: JSTokenNode) => d.width/2)
-          .attr("y", 10)
-          .attr("dy", ".35em")
-          .attr("text-anchor", "middle")
-          .style("fill", "black")
+        .attr("transform", (d: JSTokenNode, i: Int) => s"translate(${d.offSet},2)")
+
+      selection
+        .append("rect")
+        .attr("rx", 6)
+        .attr("ry", 6)
+        .attr("width", (d: JSTokenNode) => d.width)
+        .attr("height", 20)
+        .style("fill", "white")
+        .style("stroke", "gray")
+
+      selection
+        .append("text")
+        .text((d: JSTokenNode) => d.token)
+        .attr("x", (d: JSTokenNode) => d.width/2)
+        .attr("y", 10)
+        .attr("dy", ".35em")
+        .attr("text-anchor", "middle")
+        .style("fill", "black")
+
+      for(i <- edges.indices){
+        var indices = (edges(i).source.position, edges(i).target)
+        if(indices._1 > indices._2)
+          indices = indices.swap
+        for(j <- 0 until i){
+          val otherIndices = (edges(j).source.position, edges(j).target)
+          if((otherIndices._1 > indices._1 && otherIndices._1 < indices._2) || (otherIndices._2 > indices._1 && otherIndices._2 < indices._2)){
+            if(edges(i).layer <= edges(j).layer){
+              edges(i).layer = edges(j).layer + 1
+            }
+          }
+        }
+      }
```

Comment:
So this is an anti-pattern in scala. For comprehensions are common, but generally speaking, 
`for (i <- edges.indices) { ... }` can be rewritten more idiomatically as 
`edges.indices.foreach { i => ... }`

The same goes for 
`for (j <- 0 until i) { ... }` could be `(0 until i).foreach { j => ... )`

Please change accordingly.